clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/ContentNodeStats.java

```diff
@@ -3,149 +3,130 @@
+            return Objects.hash(bucketsTotal, bucketsPending);
         }
 
+        @Override
         public String toString() {
-            return String.format("{buckets = %d}", buckets);
+            return "{bucketsTotal=" + bucketsTotal + ", bucketsPending=" + bucketsPending + "}";
         }
     }
 
-    private final Amount syncing;
-    private final Amount copyingIn;
-    private final Amount movingOut;
-    private final Amount copyingOut;
-    private int nodeIndex;
+    public ContentNodeStats(StorageNode storageNode) {
+        this.nodeIndex = storageNode.getIndex();
+        for (StorageNode.BucketSpaceStats stats : storageNode.getBucketSpacesStats()) {
+            if (stats.valid()) {
+                this.bucketSpaces.put(stats.getName(),
+                        new BucketSpaceStats(stats.getBucketStats().getTotal(),
+                                stats.getBucketStats().getPending()));
+            } else {
+                // TODO: better handling of invalid bucket space stats
+                this.bucketSpaces.put(stats.getName(), new BucketSpaceStats());
+            }
+        }
+    }
 
-    /**
-     * An instance with all 0 amounts.
-     */
     public ContentNodeStats(int index) {
-        this(index, new Amount(), new Amount(), new Amount(), new Amount());
+        this(index, new HashMap<>());
     }
 
-    ContentNodeStats(int index, Amount syncing, Amount copyingIn, Amount movingOut, Amount copyingOut) {
+    public ContentNodeStats(int index, Map<String, BucketSpaceStats> bucketSpaces) {
         this.nodeIndex = index;
-        this.syncing = syncing;
-        this.copyingIn = copyingIn;
-        this.movingOut = movingOut;
-        this.copyingOut = copyingOut;
+        this.bucketSpaces = bucketSpaces;
     }
 
-    public void set(ContentNodeStats stats) {
-        nodeIndex = stats.nodeIndex;
-        syncing.set(stats.syncing);
-        copyingIn.set(stats.copyingIn);
-        movingOut.set(stats.movingOut);
-        copyingOut.set(stats.copyingOut);
-    }
+    public int getNodeIndex() { return nodeIndex; }
 
-    int getNodeIndex() { return nodeIndex; }
-    public Amount getSyncing() { return syncing; }
-    public Amount getCopyingIn() { return copyingIn; }
-    public Amount getMovingOut() { return movingOut; }
-    public Amount getCopyingOut() { return copyingOut; }
+    public void add(ContentNodeStats stats) {
+        merge(stats, 1);
+    }
 
-    void add(ContentNodeStats stats) {
-        scaledAdd(1, stats);
+    public void subtract(ContentNodeStats stats) {
+        merge(stats, -1);
     }
 
-    void subtract(ContentNodeStats stats) {
-        scaledAdd(-1, stats);
+    private void merge(ContentNodeStats stats, int factor) {
+        for (Map.Entry<String, BucketSpaceStats> entry : stats.bucketSpaces.entrySet()) {
+            BucketSpaceStats statsToUpdate = bucketSpaces.get(entry.getKey());
+            if (statsToUpdate == null && factor == 1) {
```

Comment:
Is it technically more correct to say `&& factor > 0` here? Doesn't matter for the _current_ usages of the method, though.